crates/loro-internal/src/op.rs

```diff
@@ -576,41 +566,43 @@ impl DeltaValue for SliceRanges {
             return Err(other);
         }
 
-        self.ranges.extend(other.ranges);
-        Ok(())
+        match (&mut self.values, &other.values) {
+            (Either::Left(left_range), Either::Left(right_range))
+                if left_range.0.end == right_range.0.start =>
+            {
+                left_range.0.end = right_range.0.end;
+                Ok(())
+            }
+            _ => {
+                // If one is SliceRange and the other is LoroValue, we can't merge
+                Err(other)
+            }
+        }
     }
 
     fn take(&mut self, target_len: usize) -> Self {
-        let mut right = Self {
-            ranges: Default::default(),
-            id: self.id.inc(target_len as i32),
-        };
-
-        let right_target_len = self.length() - target_len;
-        let mut right_len = 0;
-        while right_len < right_target_len {
-            let range = self.ranges.pop().unwrap();
-            let range_len = range.content_len();
-            if right_len + range_len <= target_len {
-                right.ranges.push(range);
-                right_len += range_len;
-            } else {
-                let new_range = range.slice(right_len * 2 - right_target_len, range_len);
-                right.ranges.push(new_range);
-                self.ranges
-                    .push(range.slice(0, right_len * 2 - right_target_len));
-                right_len = right_target_len;
+        match &mut self.values {
+            Either::Left(range) => {
+                let ans = range.slice(0, target_len);
+                let this = range.slice(target_len, range.atom_len());
+                *range = this;
+                let old_id = self.id;
+                self.id = self.id.inc(target_len as i32);
+                Self {
+                    id: old_id,
+                    values: Either::Left(ans),
+                    elem_id: None,
+                }
             }
+            Either::Right(_) => unimplemented!(),
```

Comment:
`unreachable` or `unimplemented`?

Reply:
unimplemented. We cannot guarantee it's unreachable at the moment